### PR TITLE
Create a new Prettier task version 1

### PR DESCRIPTION
**Pull request by YolandaDu1997**
# Changes

Create a new Prettier task to check codes from a git repo (need to use the git-clone task to fetch codes).

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [ ] Yaml file complies with [yamllint](https://github.com/adrienverge/yamllint) rules.

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

File: prettier/prettier-check.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  - name: source
+    mountPath: /source
+  params:
+  - name: subdirectory
+    type: string
+    description: Subdirectory inside "source" workspace that contains the target files.
+    default: "."
+  - name: filepath
+    type: string
+    description: path of file of directory inside subdirectory that contains the target file(s).
```
Comment (sbwsg) [inline prettier/prettier-check.yaml:16]: Did you mean "file or directory" instead of "file of directory" here?

File: prettier/prettier-check.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: prettier-check-from-gitrepo
```
Comment (sbwsg) [inline prettier/prettier-check.yaml:4]: This Task doesn't require a git repo does it?  It looks like it'll work on any code provided in a workspace.  Suggest `prettier-check` as the Task name if that's the case, dropping the "from-gitrepo" bit.

File: prettier/prettier-check.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+spec:
```
Comment (sbwsg) [inline prettier/prettier-check.yaml:5]: I suggest adding a `description` field under the `spec` that describes in plain english what this Task (and prettier) does.  Something like the following:

```
description: Run the prettier code formatter against provided source code. See https://prettier.io for more information.
```

File: prettier/prettier-check.yaml (hunk 1 above)
Comment (YolandaDu1997) [reply prettier/prettier-check.yaml:16]: Yeah, sorry it is a typo

File: prettier/prettier-check.yaml (hunk 2 above)
Comment (YolandaDu1997) [reply prettier/prettier-check.yaml:4]: I do require a git repo. I notice that there is already a git-clone task in the catalog, which will fetch the code from the git repo. Should I add prettier-check task which acts the same as git-clone?

**Comment by YolandaDu1997** [comment]
It depends on the user's input. The default mode is just to check the code without formatting. But the user can define their own options to format the code from the parameter "options".

File: prettier/prettier-check.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  workspaces:
+  - name: source
+    mountPath: /source
+  params:
+  - name: subdirectory
+    type: string
+    description: Subdirectory inside "source" workspace that contains the target files.
+    default: "."
+  - name: filepath
```
Comment (wlynch) [inline prettier/prettier-check.yaml:14]: Looking at the docs, it looks like you can provide multiple paths:

```
prettier [options] [file/dir/glob ...]
```

We may want to make some small changes to allow multiple paths:
1. rename `options` to `args`, default it to `--check .`
2. set `workingDir` to the workspace path (this way all paths can be relative to the workspace)

File: prettier/prettier-check.yaml (hunk 2 above)
Comment (wlynch) [reply prettier/prettier-check.yaml:4]: The cool thing about workspaces is that you don't need to know if it's a git repo or not! :D 
You could also have a workspace that's populated by a GCS bucket, a hg/svn repo, or anything else!

All this task needs to care about is that some filesystem is being passed to it.

**Comment by chmouel** [comment]
Thanks @YolandaDu1997 for contributing this, is there any chance if you can look if it's not too difficult to add a test for this ? there is some examples in the repository and some [documentation](https://github.com/tektoncd/catalog/blob/v1beta1/CONTRIBUTING.md#end-to-end-testing) that should get you started...

File: prettier/prettier-check.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: prettier-check
```
Comment (wlynch) [inline prettier/prettier-check.yaml:4]: nit: rename this from `prettier-check`  -> `prettier` since this might do more than check.

File: prettier/examples/prettier-format-from-gitrepo.yaml [hunk 6]
```diff
@@ -0,0 +1,54 @@
+kind: PersistentVolumeClaim
```
Comment (wlynch) [inline prettier/examples/prettier-format-from-gitrepo.yaml:1]: If you move this file from `examples` -> `tests` this will also double as e2e tests! :D

File: prettier/examples/prettier-format-from-gitrepo.yaml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  tasks:
+  - name: fetch-repo
+    taskRef:
+      name: git-clone
+    workspaces:
+    - name: output
+      workspace: shared-workspace
+    params:
+    - name: url
+      value: https://github.com/tektoncd/pipeline.git
```
Comment (wlynch) [inline prettier/examples/prettier-format-from-gitrepo.yaml:29]: Let's pin the revision fetched to the SHA of this commit. We don't want this test breaking without an explicit change.

File: prettier/README.md [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+This `Task` is for running [Prettier](https://prettier.io/) code formatter against the provided source code in the Workspace. By default this `Task` will check the format of files in the root of your Workspace.
+
+## Workspaces
+
+ - **source**: A workspace for this Task to fetch the source the code. 
+ 
+## Parameters
+
+ - **args**: Prettier arguments used to run Prettier. See [Prettier CLI](https://prettier.io/docs/en/cli.html) for the arguments. (_default:_ --check .)
```
Comment (wlynch) [inline prettier/README.md:11]: nit: 
```
`--check .`
```

File: prettier/examples/prettier-format-from-gitrepo.yaml (hunk 7 above)
Comment (YolandaDu1997) [reply prettier/examples/prettier-format-from-gitrepo.yaml:29]: I'm not sure what you mean. Does this mean that this way would probably lead to this test breaking?

Comment (wlynch) [reply prettier/examples/prettier-format-from-gitrepo.yaml]: Talked about this more in chat. Intent here is to pin the SHA being used so that an unexpected change to master doesn't cause this test to suddenly fail.

**Comment by YolandaDu1997** [comment]
> I don't think we should include a copy of the git-clone task here. We should be able to use a [`pre-apply-task-hook.sh`](https://github.com/tektoncd/catalog/blob/v1beta1/CONTRIBUTING.md#end-to-end-testing) script to install the git-clone task remotely.
> 
> e.g.
> 
> ```shell
> kubectl apply -f https://raw.githubusercontent.com/tektoncd/catalog/bf7e538778ce6c1ea8ebc72bd0d050362e63a716/git/git-clone.yaml
> ```

Sure. I am working on it.

**Comment by wlynch** [comment]
Looks good! :D 

The only thing left to do is squash all the commits down to 1 commit so the commit history is clean.

Thanks! :)

**Comment by YolandaDu1997** [comment]
> Looks good! :D
> 
> The only thing left to do is squash all the commits down to 1 commit so the commit history is clean.
> 
> Thanks! :)

Sure!

File: prettier/prettier.yaml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  params:
+  - name: args
+    type: array
+    description: Prettier arguments used to run Prettier.
+    default:
+      - "--check"
+      - "."
+  steps:
+  - name: check
+    image: elnebuloso/prettier
```
Comment (dibyom) [inline prettier/prettier.yaml:19]: v minor thing: We are not very consistent about doing this across our tasks but some tasks allow a user to specify the container image as a param (with a default set). This allows the user to use the same task but with say a different image (say, newer version of prettier)

File: prettier/tests/pre-apply-task-hook.sh [hunk 10]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+# Add git-clone
+kubectl -n ${tns} apply -f https://raw.githubusercontent.com/tektoncd/catalog/bf7e538778ce6c1ea8ebc72bd0d050362e63a716/git/git-clone.yaml
```
Comment (dibyom) [inline prettier/tests/pre-apply-task-hook.sh:4]: very cool! Did not know the test framework allowed this!

File: prettier/tests/run.yaml [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    runAfter:
+    - fetch-repo
+    workspaces:
+    - name: source
+      workspace: shared-workspace
+    params: 
+    - name: args
+      value: 
+        - --write 
+        - "config/config-defaults.yaml"
```
Comment (dibyom) [inline prettier/tests/run.yaml:33]: curious, is this default config coming from the container image? and users are able to specify a prettier config file from the workspace right?

File: prettier/prettier.yaml (hunk 9 above)
Comment (YolandaDu1997) [reply prettier/prettier.yaml:19]: Should I also make the image as a parameter and make this `elnebuloso/prettier` as the default image?

File: prettier/tests/run.yaml (hunk 11 above)
Comment (YolandaDu1997) [reply prettier/tests/run.yaml:33]: This is the file from the git repo used by `prettier` to format. It acts as a test file to be formatted in the repository.

File: prettier/prettier.yaml (hunk 9 above)
Comment (dibyom) [reply prettier/prettier.yaml:19]: yeah, that'd be nice. We can set the default to one of the versions from here: https://hub.docker.com/r/elnebuloso/prettier/tags

File: prettier/tests/run.yaml (hunk 11 above)
Comment (dibyom) [reply prettier/tests/run.yaml:33]: ahh! ignore my comment! I was thinking of the [prettier config file](https://prettier.io/docs/en/configuration.html) which allows the user to set their linter settings...but it looks like prettier would auto-detect the file based on file naming convention.

File: prettier/prettier.yaml (hunk 9 above)
Comment (YolandaDu1997) [reply prettier/prettier.yaml]: Sure!

File: prettier/tests/pre-apply-task-hook.sh (hunk 10 above)
Comment (chmouel) [reply prettier/tests/pre-apply-task-hook.sh:4]: what is the issue doing it directly from the current repository ? like the other task are doing it ? i.e: 

https://github.com/tektoncd/catalog/blob/v1beta1/buildah/tests/pre-apply-task-hook.sh#L7-L8

File: prettier/tests/run.yaml [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+  tasks:
+  - name: fetch-repo
+    taskRef:
+      name: git-clone
+    workspaces:
+    - name: output
+      workspace: shared-workspace
+    params:
+    - name: url
+      value: https://github.com/tektoncd/pipeline.git
```
Comment (chmouel) [inline prettier/tests/run.yaml:18]: I haven't checked how long this run, but should we maybe use a repository that's not as big as `pipeline`? The current catalog tests can get a long time to run and using a smaller git repository may help, i.e: something like a helloworld https://github.com/go-training/helloworld repository?

**Comment by chmouel** [comment]
A few suggestions about the tests and we are almost there for merging 😅

It would be great if you could squash all the commits before we can merge it so we can have a clean history.

**Comment by chmouel** [comment]
Looks good to me, but i'd like to know about if it was a requirements i have missed to have to do remote install of the git-clone task and not installing the git-clone task directly from the local repo, i.e: this comment : 

https://github.com/tektoncd/catalog/pull/351/files#diff-
ddbbbde00b873c8899f4e20314f26fc6R4

we may end up with an old git-clone in the future...

**Comment by YolandaDu1997** [comment]
> Looks good to me, but i'd like to know about if it was a requirements i have missed to have to do remote install of the git-clone task and not installing the git-clone task directly from the local repo, i.e: this comment :
> 
> https://github.com/tektoncd/catalog/pull/351/files#diff-ddbbbde00b873c8899f4e20314f26fc6R4
> 
> we may end up with an old git-clone in the future...

Sorry, I left it out yesterday. I'll fix it.

File: prettier/tests/run.yaml (hunk 12 above)
Comment (wlynch) [reply prettier/tests/run.yaml]: I would prefer to not depend on an external repo. If/when they make a change to their repo, it might break our tests downstream.

Since it looks like the catalog repo is available for tests anyway, how about we run prettier over the catalog? It has an added benefit that we'll now have a catalog tasks to ensure consistent formatting! 😄

Comment (YolandaDu1997) [reply prettier/tests/run.yaml]: That makes sense. Should I use the commit SHA from the `master` branch or `v1beta1` branch or both okay?

Comment (wlynch) [reply prettier/tests/run.yaml]: I think we may not even need the git-clone task.
Based on https://github.com/tektoncd/catalog/pull/351#issuecomment-642575792 it sounds like the catalog repo is already available to running tests by default? So we may just be able to run the linter on the local path.

You'd have to play around with the test environment to find out for sure.

**Comment by YolandaDu1997** [comment]
> Could you squash the commits to one? Otherwise good to go!



> Could you squash the commits to one? Otherwise good to go!

Done！

**Comment by chmouel** [comment]
/approve

thank you,

(i don't think we need to pin the catalog to a sha)

**Comment by chmouel** [comment]
maven task is failing because of a repo rename from "master" to "main" 🤦 


```
+ /ko-app/git-init -url https://github.com/spring-projects/spring-petclinic -revision master -refspec  -path /workspace/output/ '-sslVerify=true' '-submodules=true' -depth 1
{"level":"error","ts":1592287842.688938,"caller":"git/git.go:41","msg":"Error running git [fetch --recurse-submodules=yes --depth=1 origin --update-head-ok --force master]: exit status 128\nfatal: couldn't find remote ref master\n","stacktrace":"github.com/tektoncd/pipeline/pkg/git.run\n\tgithub.com/tektoncd/pipeline/pkg/git/git.go:41\ngithub.com/tektoncd/pipeline/pkg/git.Fetch\n\tgithub.com/tektoncd/pipeline/pkg/git/git.go:116\nmain.main\n\tgithub.com/tektoncd/pipeline/cmd/git-init/main.go:53\nruntime.main\n\truntime/proc.go:203"}
{"level":"fatal","ts":1592287842.6890328,"caller":"git-init/main.go:54","msg":"Error fetching git repository: failed to fetch [master]: exit status 128","stacktrace":"main.main\n\tgithub.com/tektoncd/pipeline/cmd/git-init/main.go:54\nruntime.main\n\truntime/proc.go:203"}
```

Let's merge https://github.com/tektoncd/catalog/pull/365 first then,
